File: frontend/src/app/model/ws-message.ts
export interface WsMessageParam {
  type: string;
  roomId?: string;
  data?: any;
  sender?: string;
  direct?: boolean;
  to?: string;
}

export class WsMessage {
  type: string;
  roomId: string;
  data: any;
  sender: string;
  direct: boolean;
  to: string;

  constructor(param: WsMessageParam) {
    if (param) {
      this.type = param.type;
      this.roomId = param.roomId;
      this.data = param.data;
      this.sender = param.sender;
      this.direct = param.direct;
      this.to = param.to;
    }
  }
}

File: backend/src/main/kotlin/com/hiper2d/util/WsMessage.kt
package com.hiper2d.util

data class WsMessage(
    val type: String,
    val data: Any?
)

data class WsRoomMessage(
    val type: String,
    val roomId: String,
    val data: Any?,
    val sender: String,
    val direct: Boolean = false,
    val to: String = ""
)
